Occurs-checker optimization

liesnikov opened the pull request:
At the moment we do occurs-check twice on the arguments to the meta, once on line 525 and if that fails, we prune and check again:
https://github.com/agda/agda/blob/a0c0e586d2cbf9a2bc8ac3597fbf6733c98fc8f7/src/full/Agda/TypeChecking/MetaVars/Occurs.hs#L525-L545

This change introduces eager pruning and does occurs check afterwards only when pruning didn't manage to remove all offending variables.

The test suite passes.
Unless something has gone wrong with the tests and the reported number is off, we can get a (significant?) speedup.
std-lib test time goes from 12min22s to 10min44s. Cubical goes up from 28min16s to 27min2s.

<details> 
  <summary>std-lib test summary <b>without</b> changes, total "real time" is 12min </summary>
  
```
Total                        743,043ms            
Miscellaneous                  6,703ms            
Typing                        41,847ms (313,277ms)
Typing.CheckRHS              101,892ms            
Typing.CheckLHS               25,035ms  (80,181ms)
Typing.CheckLHS.UnifyIndices  55,145ms            
Typing.Generalize             27,722ms            
Typing.TypeSig                26,037ms            
Typing.OccursCheck            22,431ms            
Typing.With                   11,342ms            
Typing.InstanceSearch          1,821ms            
Coverage                     112,918ms            
Serialization                 87,394ms (105,123ms)
Serialization.Compress         8,493ms            
Serialization.BinaryEncode     6,498ms            
Serialization.Sort             1,925ms            
Serialization.BuildInterface     811ms            
DeadCode                      46,651ms            
Parsing                        5,343ms  (41,749ms)
Parsing.OperatorsExpr         30,028ms            
Parsing.OperatorsPattern       6,376ms            
Deserialization               29,869ms  (40,962ms)
Deserialization.Compaction    11,092ms            
Scoping                        8,282ms  (34,770ms)
Scoping.InverseScopeLookup    26,487ms            
Positivity                    26,094ms            
Termination                    2,370ms   (7,063ms)
Termination.RecCheck           3,518ms            
Termination.Compare              829ms            
Termination.Graph                344ms            
Highlighting                   4,792ms            
Injectivity                    1,412ms            
Import                         1,270ms            
ProjectionLikeness               212ms            
ModuleName                        40ms            
 918,679,447,304 bytes allocated in the heap
 103,950,975,624 bytes copied during GC
   1,545,617,576 bytes maximum residency (163 sample(s))
       4,712,144 bytes maximum slop
            2256 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     218520 colls,     0 par   148.770s  149.016s     0.0007s    0.0435s
  Gen  1       163 colls,     0 par   21.507s  21.509s     0.1320s    0.9957s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.000s elapsed)
  MUT     time  572.902s  (571.633s elapsed)
  GC      time  170.277s  (170.526s elapsed)
  EXIT    time    0.001s  (  0.001s elapsed)
  Total   time  743.180s  (742.160s elapsed)

  Alloc rate    1,603,555,129 bytes per MUT second

  Productivity  77.1% of total user, 77.0% of total elapsed


real	12m22,337s
user	12m15,298s
sys	0m8,060s
  ```
  
</details>

<details> 
  <summary>std-lib test summary <b>with</b> changes, total "real time" is 10min44s </summary>

```
Total                        644,843ms            
Miscellaneous                  5,616ms            
Typing                        36,933ms (270,357ms)
Typing.CheckRHS               82,973ms            
Typing.CheckLHS               20,895ms  (72,077ms)
Typing.CheckLHS.UnifyIndices  51,182ms            
Typing.Generalize             26,403ms            
Typing.TypeSig                22,605ms            
Typing.OccursCheck            19,254ms            
Typing.With                    8,317ms            
Typing.InstanceSearch          1,793ms            
Coverage                     100,873ms            
Serialization                 72,565ms  (89,835ms)
Serialization.Compress         7,944ms            
Serialization.BinaryEncode     7,002ms            
Serialization.Sort             1,620ms            
Serialization.BuildInterface     702ms            
DeadCode                      43,504ms            
Parsing                        4,483ms  (34,415ms)
Parsing.OperatorsExpr         24,855ms            
Parsing.OperatorsPattern       5,076ms            
Deserialization               23,416ms  (32,723ms)
Deserialization.Compaction     9,307ms            
Scoping                        7,029ms  (31,732ms)
Scoping.InverseScopeLookup    24,702ms            
Positivity                    23,874ms            
Termination                    1,919ms   (5,939ms)
Termination.RecCheck           3,069ms            
Termination.Compare              674ms            
Termination.Graph                275ms            
Highlighting                   3,641ms            
Injectivity                    1,223ms            
Import                           900ms            
ProjectionLikeness               181ms            
ModuleName                        21ms            
 918,836,033,880 bytes allocated in the heap
 103,728,040,952 bytes copied during GC
   1,545,894,712 bytes maximum residency (166 sample(s))
       5,105,040 bytes maximum slop
            2255 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     218551 colls,     0 par   128.029s  128.206s     0.0006s    0.0371s
  Gen  1       166 colls,     0 par   20.849s  20.851s     0.1256s    1.0150s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time  496.091s  (495.418s elapsed)
  GC      time  148.878s  (149.056s elapsed)
  EXIT    time    0.001s  (  0.005s elapsed)
  Total   time  644.971s  (644.480s elapsed)

  Alloc rate    1,852,151,892 bytes per MUT second

  Productivity  76.9% of total user, 76.9% of total elapsed


real	10m44,650s
user	10m38,482s
sys	0m6,659s
```
</details>

<details> 
  <summary>cubical test summary <b>without</b> changes, total "real time" is 28min16s </summary>
  
```
———— All done; warnings encountered ————————————————————————

/home/bohdan/delft/agda/agda/cubical/Cubical/Tactics/Reflection.agda:92,5-19
warning: -W[no]UserWarning
DEPRECATED: Use `withReduceDefs` instead of `dontReduceDefs`
when scope checking dontReduceDefs
make[2]: Leaving directory '/home/bohdan/delft/agda/agda/cubical'
make[1]: Leaving directory '/home/bohdan/delft/agda/agda/cubical'

real	28m16,383s
user	28m13,232s
sys	0m4,640s
```
  
</details>

<details> 
  <summary>cubical test summary <b>with</b> changes, total "real time" is 27min2s </summary>
  
```
———— All done; warnings encountered ————————————————————————

/home/bohdan/delft/agda/agda/cubical/Cubical/Tactics/Reflection.agda:92,5-19
warning: -W[no]UserWarning
DEPRECATED: Use `withReduceDefs` instead of `dontReduceDefs`
when scope checking dontReduceDefs
make[2]: Leaving directory '/home/bohdan/delft/agda/agda/cubical'
make[1]: Leaving directory '/home/bohdan/delft/agda/agda/cubical'

real	27m2,328s
user	26m59,405s
sys	0m4,051s
```

</details>

liesnikov committed:
run occurs-check only once on metas

liesnikov committed:
don't roll back the state after some pruning

and don't do occurs-check after full pruning

jespercockx reviewed (approved):
Thanks for the PR! I'll run this on my machine as well to see if I can reproduce the performance improvements.

One very small nitpick: could you swap the use of `Left` and `Right`? It makes more sense to me that "successfully pruned everything" would be the success case.

liesnikov committed:
add comments

liesnikov committed:
swap left and right per comment

https://github.com/agda/agda/pull/6900#pullrequestreview-1661328042

liesnikov committed:
restore issue reference in a comment

jespercockx commented:
I don't seem to get any speedup on my machine. Did you compile both times with `--optimize-heavily`?

liesnikov commented:
Yeah, the `make install-bin`.

But it not making a difference makes more sense.

liesnikov commented:
@andreasabel I was talking with @jespercockx and we believe it still makes sense to merge this just for code simplification purposes.
Could you perhaps take a look and let us know if you approve?

andreasabel commented on src/full/Agda/TypeChecking/MetaVars/Occurs.hs:
```diff
@@ -521,30 +521,31 @@ instance Occurs Term where
+            t <- caseMaybe (allApplyElims es) (fallback) $ \ vs -> do
+              -- don't attempt pruning if we're in a flexible context
+               if not (isFlexible ctx) then do
+                 killResult <- lift . prune m' vs =<< allowedVars
```
It seems that the new occurs check invokes pruning eagerly, as long as the conditions are met (rigid context, no projections on the meta).
The old check invoked pruning only when there was a pattern violation when checking the meta arguments.

Attempting pruning is never wrong, but might not produce any pruning, if the meta did not have a bad argument.  This is why pruning was only attempted when there was reason to believe it could succeed.

As it stands I do not see this PR as an _optimization_ of the occurs check; rather, a change of strategy.
As such, there should be some evaluation of old vs. new strategy.

andreasabel commented on src/full/Agda/TypeChecking/MetaVars/Occurs.hs:
```diff
@@ -521,30 +521,31 @@ instance Occurs Term where
+              -- Sometimes instantiated meta is actually not a meta itself
+              (Left t) -> occurs t
+              -- the PrunedEverything case
+              (Right v) -> return v
```
The old occurs check would restart on the pruned `v`, the new one does not.  Is it clear that the result of `PrunedEveryting` is always passing the occurs check?  If yes, it should be documented as postcondition in the `prune` function.

andreasabel reviewed (changes requested):
If I understand correctly, this is strategy change in the occurs checker.
Could the PR include evidence that the new strategy is better than the old?
Also, there should be more documentation of its soundness (as `occurs` isn't rerun on the pruned term).
(See detailed comments below.)

liesnikov commented:
Thank you!

I titled the PR "optimization" because it seemed to run faster, but apparently that was a fluke.
I'll see what do we get with cost centre analysis.

liesnikov committed:
spelling

liesnikov committed:
don't instantiate if we didn't manage to prune anything

and don't do occurs check if we didn't need to prune anything in the
first place

liesnikov commented:
A small status update:
Profiling says the change actually causes a slight slowdown on (parts of) std-lib before 9733af7 and  9733af7 itself seemingly introduces a space leak, exhausting the heap on Data.Nat.Induction no matter how much is provided.

In cases when it doesn't fail 9733af seems to type-check insignificantly faster than current master, but all other phases experience similarly insignificant slowdown, so the net effect is about zero.

I'm not sure what's the cause for the space leak, perhaps the fact that we don't instantiate in some cases?
I might investigate it later, but for now I think it's sensible to close the PR.

<details> 
  <summary>profiling summary</summary>


`agda-main-prof` is https://github.com/liesnikov/agda/tree/bench-pruning
`agda-prof` is the occurs-twice branch but with commits adding pruning phase cherry-picked from the bench-pruning branch

`with nothingtoprune case` corresponds to commit 9733af7

```
profiling internal on Algebra with proper billing

std-lib ⊢agda-main-prof --profile=internal -i. -isrc src/Algebra.agda


Total                        30,145ms           
Miscellaneous                   169ms           
Typing                        5,535ms (11,758ms)
Typing.CheckRHS               2,609ms           
Typing.Generalize             1,363ms           
Typing.TypeSig                  956ms           
Typing.CheckLHS                 635ms    (709ms)
Typing.CheckLHS.UnifyIndices     74ms           
Typing.OccursCheck              340ms    (464ms)
Typing.OccursCheck.Pruning      123ms           
Typing.With                     117ms           
Serialization                 7,230ms  (7,750ms)
Serialization.BinaryEncode      213ms           
Serialization.Compress          195ms           
Serialization.Sort               78ms           
Serialization.BuildInterface     32ms           
DeadCode                      2,501ms           
Positivity                    2,439ms           
Parsing                         888ms  (2,143ms)
Parsing.OperatorsExpr         1,134ms           
Parsing.OperatorsPattern        120ms           
Deserialization                 999ms  (1,183ms)
Deserialization.Compaction      184ms           
Coverage                        684ms           
Scoping                         499ms    (658ms)
Scoping.InverseScopeLookup      159ms           
Highlighting                    424ms           
Termination                      22ms    (277ms)
Termination.RecCheck            255ms           
Import                           98ms           
Injectivity                      37ms           
ProjectionLikeness               16ms           


std-lib ⊢agda-prof --profile=internal -i. -isrc src/Algebra.agda

Total                        31,698ms           
Miscellaneous                   170ms           
Typing                        5,635ms (12,106ms)
Typing.CheckRHS               2,683ms           
Typing.Generalize             1,346ms           
Typing.TypeSig                  988ms           
Typing.CheckLHS                 688ms    (765ms)
Typing.CheckLHS.UnifyIndices     76ms           
Typing.OccursCheck              346ms    (555ms)
Typing.OccursCheck.Pruning      208ms           
Typing.With                     130ms           
Serialization                 7,838ms  (8,599ms)
Serialization.BinaryEncode      438ms           
Serialization.Compress          202ms           
Serialization.Sort               85ms           
Serialization.BuildInterface     33ms           
DeadCode                      2,670ms           
Positivity                    2,617ms           
Parsing                         911ms  (2,197ms)
Parsing.OperatorsExpr         1,162ms           
Parsing.OperatorsPattern        123ms           
Deserialization                 909ms  (1,102ms)
Deserialization.Compaction      193ms           
Coverage                        692ms           
Scoping                         524ms    (655ms)
Scoping.InverseScopeLookup      130ms           
Highlighting                    402ms           
Termination                      23ms    (309ms)
Termination.RecCheck            285ms           
Import                          118ms           
Injectivity                      38ms           
ProjectionLikeness               17ms           


with nothingtoprune case

Total                        30,122ms           
Miscellaneous                   166ms           
Typing                        5,576ms (11,571ms)
Typing.CheckRHS               2,541ms           
Typing.Generalize             1,267ms           
Typing.TypeSig                  884ms           
Typing.CheckLHS                 655ms    (727ms)
Typing.CheckLHS.UnifyIndices     72ms           
Typing.OccursCheck              302ms    (468ms)
Typing.OccursCheck.Pruning      165ms           
Typing.With                     105ms           
Serialization                 7,456ms  (8,215ms)
Serialization.BinaryEncode      426ms           
Serialization.Compress          197ms           
Serialization.Sort               79ms           
Serialization.BuildInterface     55ms           
DeadCode                      2,449ms           
Positivity                    2,425ms           
Parsing                         913ms  (2,155ms)
Parsing.OperatorsExpr         1,108ms           
Parsing.OperatorsPattern        134ms           
Deserialization                 859ms  (1,045ms)
Deserialization.Compaction      186ms           
Coverage                        640ms           
Scoping                         481ms    (609ms)
Scoping.InverseScopeLookup      128ms           
Highlighting                    404ms           
Termination                      21ms    (288ms)
Termination.RecCheck            266ms           
Import                           96ms           
Injectivity                      36ms           
ProjectionLikeness               16ms           

----

profiling internal on README/Data.agda

agda-prof --profile=internal -i. -isrc README/Data.agda

Total                        665,840ms            
Miscellaneous                  2,514ms            
Typing                        37,835ms (263,381ms)
Typing.CheckRHS               76,056ms            
Typing.CheckLHS               22,519ms  (69,212ms)
Typing.CheckLHS.UnifyIndices  46,692ms            
Typing.Generalize             29,696ms            
Typing.TypeSig                20,767ms            
Typing.OccursCheck             8,245ms  (20,614ms)
Typing.OccursCheck.Pruning    12,368ms            
Typing.With                    8,225ms            
Typing.InstanceSearch            971ms            
Serialization                124,094ms (135,284ms)
Serialization.BinaryEncode     5,489ms            
Serialization.Compress         3,209ms            
Serialization.Sort             1,954ms            
Serialization.BuildInterface     536ms            
Coverage                      93,701ms            
Parsing                       11,183ms  (45,398ms)
Parsing.OperatorsExpr         28,271ms            
Parsing.OperatorsPattern       5,943ms            
DeadCode                      44,714ms            
Scoping                        6,165ms  (28,929ms)
Scoping.InverseScopeLookup    22,764ms            
Positivity                    20,118ms            
Deserialization               14,701ms  (18,474ms)
Deserialization.Compaction     3,773ms            
Termination                    1,965ms   (6,731ms)
Termination.RecCheck           3,813ms            
Termination.Compare              759ms            
Termination.Graph                193ms            
Highlighting                   4,209ms            
Injectivity                    1,374ms            
Import                           740ms            
ProjectionLikeness               248ms            
ModuleName                        16ms            

agda-main-prof --profile=internal -i. -isrc README/Data.agda

Total                        623,199ms            
Miscellaneous                  2,336ms            
Typing                        36,634ms (246,890ms)
Typing.CheckRHS               71,587ms            
Typing.CheckLHS               21,204ms  (65,109ms)
Typing.CheckLHS.UnifyIndices  43,905ms            
Typing.Generalize             25,623ms            
Typing.TypeSig                19,628ms            
Typing.OccursCheck             7,809ms  (18,928ms)
Typing.OccursCheck.Pruning    11,119ms            
Typing.With                    8,476ms            
Typing.InstanceSearch            902ms            
Serialization                116,070ms (126,988ms)
Serialization.BinaryEncode     5,708ms            
Serialization.Compress         3,104ms            
Serialization.Sort             1,607ms            
Serialization.BuildInterface     497ms            
Coverage                      88,433ms            
DeadCode                      45,109ms            
Parsing                       10,059ms  (41,310ms)
Parsing.OperatorsExpr         25,835ms            
Parsing.OperatorsPattern       5,414ms            
Scoping                        5,575ms  (26,490ms)
Scoping.InverseScopeLookup    20,914ms            
Positivity                    19,155ms            
Deserialization               10,847ms  (14,212ms)
Deserialization.Compaction     3,364ms            
Termination                    1,720ms   (6,267ms)
Termination.RecCheck           3,633ms            
Termination.Compare              734ms            
Termination.Graph                179ms            
Highlighting                   3,742ms            
Injectivity                    1,273ms            
Import                           747ms            
ProjectionLikeness               228ms            
ModuleName                        12ms            
```  
  
 </details>

andreasabel commented:
@liesnikov Thanks for your investigation on this issue.
I feel sorry that nothing actionable came out of this now, but maybe this the harsh reality of type-checker performance tuning: It is hard to get conclusive results.
